File: app/src/main/java/com/lence/penguinsvskillerwhales/life/PrimaryState.java
package com.lence.penguinsvskillerwhales.life;

import com.lence.penguinsvskillerwhales.model.Orca;
import com.lence.penguinsvskillerwhales.model.Penguin;
import com.lence.penguinsvskillerwhales.view.KotlinField;
import com.lence.penguinsvskillerwhales.view.UpdateAdapter;
import com.lence.penguinsvskillerwhales.view.UpdatePresenter;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Random;
import java.util.Set;

public class PrimaryState {
    private int mRows;
    private int mColumns;
    private UpdatePresenter mUpdatePresenter;
    //private List<Organism> mLists;
    private KotlinField mField;
    private UpdateAdapter mUpdateAdapter;

    PrimaryState(int rows, int columns, UpdatePresenter updatePresenter, UpdateAdapter updateAdapter, KotlinField field) {

        mRows = rows;
        mColumns = columns;
        mUpdatePresenter = updatePresenter;
        mUpdateAdapter = updateAdapter;
        mField = field;
    }

    public void execute() {
        //Field.setmLists(Collections.synchronizedList(new ArrayList<Organism>()));
        mField = new KotlinField(mRows, mColumns);
        if (mRows * mColumns <= 1) {
            mField.add(new Penguin(mUpdatePresenter));
        } else if (mRows * mColumns == 2) {
            mField.add(new Penguin(mUpdatePresenter));
            mField.add(new Orca(mUpdatePresenter));
        } else {
            for (int y = 0; y < mRows; y++)
                for (int x = 0; x < mColumns; x++) {
                    mField.add(null);
                }
            Set<Integer> generated = new HashSet<>();
            int generatedCount = (mRows * mColumns) / 2;
            if (mRows * mColumns / 20 > 0) {
                generatedCount += ((mRows * mColumns) / 20);
            } else {
                //if table is small then add 1 orcas
                generatedCount += 1;
            }
            Random random = new Random();
            while (generated.size() < generatedCount) {
                generated.add(random.nextInt((mRows * mColumns) - 1));
            }
            Iterator<Integer> iterator = generated.iterator();
            int generatedIndex = 0;
            for (; generatedIndex < (mRows * mColumns) / 2; generatedIndex++) {
                int position = iterator.next();
                mField.set( position % mColumns,position / mColumns, new Penguin(mUpdatePresenter));
            }
            for (; generatedIndex < generatedCount; generatedIndex++) {
                int position = iterator.next();
                mField.set( position % mColumns,position / mColumns, new Orca(mUpdatePresenter));
            }
        }
        mUpdatePresenter.update(mField);
        mUpdateAdapter.createAdapter(mField);
    }

}

File: app/src/main/java/com/lence/penguinsvskillerwhales/life/PrimaryState.kt
package com.lence.penguinsvskillerwhales.life

import com.lence.penguinsvskillerwhales.model.Orca
import com.lence.penguinsvskillerwhales.model.Penguin
import com.lence.penguinsvskillerwhales.view.Field
import com.lence.penguinsvskillerwhales.view.UpdateAdapter
import com.lence.penguinsvskillerwhales.view.UpdatePresenter
import java.util.*

class PrimaryState internal constructor(private val mRows: Int,
                                        private val mColumns: Int,
                                        private val mUpdatePresenter: UpdatePresenter,
                                        private val mUpdateAdapter: UpdateAdapter,
                                        private var mField: Field?) {

    fun execute() {
        mField = Field(mRows, mColumns)
        if (mRows * mColumns <= 1) {
            mField!!.add(Penguin(mUpdatePresenter))
        } else if (mRows * mColumns == 2) {
            mField!!.add(Penguin(mUpdatePresenter))
            mField!!.add(Orca(mUpdatePresenter))
        } else {
            for (y in 0 until mRows)
                for (x in 0 until mColumns) {
                    mField!!.add(null)
                }
            val generated = HashSet<Int>()
            var generatedCount = mRows * mColumns / 2
            if (mRows * mColumns / 20 > 0) {
                generatedCount += mRows * mColumns / 20
            } else {
                //if table is small then add 1 orcas
                generatedCount += 1
            }
            val random = Random()
            while (generated.size < generatedCount) {
                generated.add(random.nextInt(mRows * mColumns - 1))
            }
            val iterator = generated.iterator()
            var generatedIndex = 0
            while (generatedIndex < mRows * mColumns / 2) {
                val position = iterator.next()
                mField!![position % mColumns, position / mColumns] = Penguin(mUpdatePresenter)
                generatedIndex++
            }
            while (generatedIndex < generatedCount) {
                val position = iterator.next()
                mField!![position % mColumns, position / mColumns] = Orca(mUpdatePresenter)
                generatedIndex++
            }
        }
        mUpdatePresenter.update(mField!!)
        mUpdateAdapter.createAdapter(mField!!)
    }

}
